### PR TITLE
WIP - eslint & code refactoring

### PR DESCRIPTION
**Pull request by PopGoesTheWza**

### CODE REVIEW

**Comment by PopGoesTheWza** [comment]
@koresar all done except `packages/check-compose`

- only the `package-lock.json` required by CI Travis are generated and commited to the repo
- every package has been refactored and has `use strict` at the top (there is a single exception which requires more attention, `packages/convert-class/index.js`)
- ES2016 syntax everywhere and deprecated items like `__proto__` are gone
- Interaction with objects are primarily done using `Reflect` and `Object` methods
- `packages/collision/index.js` would still benefit from some more love for `Collision` is composed using some composer-from-hell which I've not fully reduced

**Comment by PopGoesTheWza** [comment]
addresses #57 and #58

File: package.json [hunk 1]
```diff
@@ -24,15 +24,23 @@
...
-    "lint": "jshint packages/**/*.js"
+    "lint": "eslint packages/*",
+    "lint-fix": "eslint --fix packages/*"
   },
   "engines": {
     "node": ">= 8.2.1"
   },
   "devDependencies": {
+    "eslint": "^6.6.0",
+    "eslint-config-airbnb-base": "^14.0.0",
```
Comment (koresar) [inline package.json]: Here is a story.

There was once a guy (actually two guys twice) which added linting to few other projects I maintain. He added application-specific linting rules (airbnb) to a low level JS library. Then he vanished because it was too difficult for him to maintain build fails which arise from ESLint modules incompatibilities. I had to remove ESLint configuration entirely because I also had no time to maintain them.

As the result: the module stayed awesome as it way before ESLint. None of us waste time any more upgrading the modules and fixing the incompatible rules.

I believe you won't be like one of those guys. I believe you'll stay with me and continue maintaining the ESLint.

File: packages/required/index.js [hunk 2]
```diff
@@ -1,53 +1,64 @@
...
-    if (isObject(settingsValue)) {
-      var metadataKeys = Object.keys(settingsValue);
-      for (var j = 0; j < metadataKeys.length; j++) {
-        var metadataKey = metadataKeys[j];
-        var metadataValue = settingsValue[metadataKey];
-        if (metadataValue === Required || metadataValue === required) {
-          // We found one thing which have to be provided. Let's check if it exists.
-          if (!descriptor[settingsKey] || descriptor[settingsKey][metadataKey] === undefined) {
-            throw new Error('Required: There must be ' + metadataKey + ' in this stamp ' + settingsKey);
+const checkDescriptorHaveThese = (descriptor, settings) => {
```
Comment (koresar) [inline packages/required/index.js]: I believe this code is also slower than the previous one. It creates too many closures.

File: packages/it/index.js [hunk 3]
```diff
@@ -1,97 +1,99 @@
...
-var isFunction = require('@stamp/is/function');
-var merge = require('@stamp/core/merge');
-var assign = require('@stamp/core/assign');
-
-var concat = Array.prototype.concat;
-function extractFunctions() {
-  var fns = concat.apply([], arguments).filter(isFunction);
+/* eslint-disable no-use-before-define */
+
+'use strict';
```
Comment (koresar) [inline packages/it/index.js]: AFAIK `use strict` is not needed in node 8 and later.

File: packages/is/plain-object.js [hunk 4]
```diff
@@ -1,4 +1,7 @@
-module.exports = function isPlainObject(value) {
-  return Boolean(value) && typeof value === 'object' &&
-    Object.getPrototypeOf(value) === Object.prototype;
-};
+'use strict';
+
+const isPlainObject = (value) =>
+  // ??? purpose of Boolean(value)
```
Comment (koresar) [inline packages/is/plain-object.js]: Try
```js
isPlainObject(null)
```

With `Boolean` it returns `false`. Without `Boolean` it returns `null`.

**Comment by koresar** [comment]
### Here is a general feedback from my 7 years of JavaScript experience
 (and 8 years more of other stuff)

TL;DR: increase in maintenance complexity for no visible (to me) benefits.

I've seen OSS project stagnate because none could maintain it. This happens daily. To avoid that pitfall I developed a set of practices to make sure my projects survive. In general - KISS (keep it stupid simple). Thus this project had so few configuration files and additional tools/modules installed. Every new (dev) dependency is the maintenance overhead, maybe not now but in months time.

Secondly. The ESLint rules used are for **production browser bundled apps** (namely - airbnb). Whereas `stampit` source code is a **low level JS library**. 

As a hint of wrongly chosen ESLint rules - see how many times you had to add `eslint-disable`. I counted 78. Wow! So much work. But what's the benefit? To satisfy airbnb?
The ESLint rules better be chosen to satisfy our project needs.

Take this low level lib for example - Preact. Here is their entire ESLint config - https://github.com/preactjs/preact/blob/master/package.json#L42
This config more suitable for stampit as well.

=====

If you still believe that airbnb was the right choice and you would like to continue maintaining it - then I'm okay with the ESLint configuration.

=====

On the non-ESLint front - well done! That's a progress to a right direction. Thank you for the effort.

File: packages/compose/index.js [hunk 5]
```diff
@@ -1,54 +1,54 @@
...
  */
-function createFactory() {
-  return function Stamp(options) {
-    var descriptor = Stamp.compose || {};
+const createFactory = () => {
+  return function Stamp(options = {}, ...args) {
+    const descriptor = Stamp.compose || {};
     // Next line was optimized for most JS VMs. Please, be careful here!
-    var obj = {__proto__: descriptor.methods}; // jshint ignore:line
+    // let obj = { __proto__: descriptor.methods };
```
[the remark this answers is not in the document]
Comment (PopGoesTheWza) [reply packages/compose/index.js]: My understanding was `@stamp` for NodeJS only, with `stampit` being aimed for other environments like browsers.

File: package.json (hunk 1 above)
Comment (koresar) [reply package.json]: I think I've just accidentally removed your comment. You are right. This repo in node only. Sorry. I forgot.
Still, we need to make sure that object creation performance is highest possible.

File: packages/compose/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (koresar) [reply packages/compose/index.js]: You are right. This repo in node only. Sorry. I forgot.
Still, we need to make sure that object creation performance is highest possible.

[the remark this answers is not in the document]
Comment (PopGoesTheWza) [reply packages/compose/index.js]: I'll revert this change pending your feedback.

File: packages/compose/index.js [hunk 6]
```diff
@@ -1,54 +1,54 @@
...
-        var returnedValue = initializer.call(obj, options,
-          {instance: obj, stamp: Stamp, args: slice.apply(arguments)});
-        obj = returnedValue === undefined ? obj : returnedValue;
-      }
+    defineProperties(obj, descriptor.propertyDescriptors || {});
+
+    // if (!descriptor.initializers || descriptor.initializers.length === 0) return obj;
+    if (descriptor.initializers && descriptor.initializers.length > 0) {
+      const inits = descriptor.initializers;
+      inits.forEach((initializer) => {
```
[the remark this answers is not in the document]
Comment (PopGoesTheWza) [reply packages/compose/index.js]: Swithing to a `for (... of ...)` statement

File: packages/compose/index.js [hunk 7]
```diff
@@ -1,54 +1,54 @@
...
+        if (isFunction(initializer)) {
+          const returnedValue = initializer.call(obj, options, {
+            instance: obj,
+            stamp: Stamp,
+            args: [options, ...args],
```
[the remark this answers is not in the document]
Comment (PopGoesTheWza) [reply packages/compose/index.js]: Is the performance due only to the `.forEach()` closure? Comparing with original code, the only difference I see is the switch from (somewhat deprecated) `arguments` to argument spread.

File: packages/compose/index.js [hunk 8]
```diff
@@ -1,54 +1,54 @@
...
 
 /**
  * Returns a new stamp given a descriptor and a compose function implementation.
  * @param {Descriptor} [descriptor={}] The information about the object the stamp will be creating.
  * @param {Compose} composeFunction The "compose" function implementation.
  * @returns {Stamp}
  */
-function createStamp(descriptor, composeFunction) {
-  var Stamp = createFactory();
+const createStamp = (descriptor, composeFunction) => {
```
[the remark this answers is not in the document]
Comment (PopGoesTheWza) [reply packages/compose/index.js]: Took me a while to realize what "array functions" was... ;)

I started programming JS in the last century. After a decade of consulting in IT software (with little if any JS), I came back two years ago to find a very different landscape: ES5, ES6... and Typescript.

This must be why I am so prone to use the "modern" syntax as much as possible.

Not to oppose with your comment but I would raise two remarks:

- the `@stamp` source code is not a place for junior JS developers to get started
- learning arrow functions is a very important part of JS in 2019

;)

File: packages/compose/index.js [hunk 9]
```diff
@@ -60,52 +60,42 @@ function createStamp(descriptor, composeFunction) {
...
-  dstObject[propName] = dstArray;
-  for (var i = 0; i < length; i += 1) {
-    var fn = srcArray[i];
-    if (isFunction(fn) && dstArray.indexOf(fn) < 0) {
-      dstArray.push(fn);
-    }
+const concatAssignFunctions = (dstObject, srcArray, propName) => {
+  if (isArray(srcArray)) {
+    const ws = new Set(get(dstObject, propName) || []);
+    srcArray.forEach((fn) => isFunction(fn) && ws.add(fn));
```
[the remark this answers is not in the document]
Comment (PopGoesTheWza) [reply packages/compose/index.js]: switching to `for (... of ...)`

File: packages/compose/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (koresar) [reply packages/compose/index.js]: yeas, the forEach. The `arguments` seem fine.

File: packages/compose/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (koresar) [reply packages/compose/index.js]: Yeah. AFAIK for-of is as performance as for in node 12. Not sure about node 10. And it's time to deprecate node 8 since it's EOL is 1-st of Jan 2020.

File: packages/compose/index.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (koresar) [reply packages/compose/index.js]: Okay. Agree to your points. Leave to your taste. :) I'm fine.

File: package.json (hunk 1 above)
Comment (PopGoesTheWza) [reply package.json]: Now another story:

I never bothered about `eslint` for most if not all the projects I own or contribute to are Typescript based, and thus relying on `tslint`.

I only recently found out that the Typescript team has officially deprecated tslint and that the whole TS ecosystem should now go the eslint way.

So adding `eslint` to `@stamp` is both

- an opportunity for me to get familiar with it
- a necessary intermediate step if we choose to convert to Typescript

The current `eslint` setup and set of active rules should not remain so strict and extended.

File: packages/required/index.js (hunk 2 above)
Comment (PopGoesTheWza) [reply packages/required/index.js]: switching to `for (... of ...)`

**Comment by koresar** [comment]
Loving your story mate! 👍
Feel free to experiment here, I see why you need this now. Apologies if I
was somewhat grumpy.

File: packages/it/index.js (hunk 3 above)
Comment (PopGoesTheWza) [reply packages/it/index.js]: Now that you mention it, `use strict` is not necessary for node scripts and commonjs modules. I'd need to check with Typescript still.

I have to habit of always writing and running in strict mode

**Comment by koresar** [comment]
AFAIK, the places where we use "this" the script mode is still necessary.
But surely not in every way file. It's sorta was deprecated in ES6 i think.

File: packages/is/plain-object.js (hunk 4 above)
Comment (PopGoesTheWza) [reply packages/is/plain-object.js]: Ok. My questioning was originally: is `Boolean(value)` equivalent to `!!value`.
I tend to forget the "typeof null" thing, sorry.

File: packages/compose/index.js [hunk 10]
```diff
@@ -1,54 +1,55 @@
-var isArray = require('@stamp/is/array');
-var isFunction = require('@stamp/is/function');
-var isObject = require('@stamp/is/object');
-var isStamp = require('@stamp/is/stamp');
-var isComposable = require('@stamp/is/composable');
+'use strict';
 
-var assign = require('@stamp/core/assign');
-var merge = require('@stamp/core/merge');
+const { isArray, isComposable, isFunction, isObject, isStamp } = require('@stamp/is');
```
Comment (koresar) [inline packages/compose/index.js]: 👍

File: packages/compose/index.js [hunk 11]
```diff
@@ -127,37 +120,32 @@ function mergeComposable(dstDescriptor, srcComposable) {
...
-  for (var i = 0; i < arguments.length; i++) {
-    var arg = arguments[i];
-    if (isComposable(arg)) {
-      mergeComposable(descriptor, arg);
-      composables.push(arg);
-    }
-  }
+  if (isComposable(this)) add(this);
+
+  args.forEach((arg) => isComposable(arg) && add(arg));
```
Comment (koresar) [inline packages/compose/index.js]: Can we please avoid `Array.prototype.forEach` altogether? I find `for-of` for readable. But most important - `for-of` is much more performant.

I even have a separate project where I measured things under multiple Node versions: https://github.com/koresar/bench-node

It clearly pointed me (at that time) that `.filter()` and `.map()` are much slower than `for` loop. The `for-of` is okay too.

File: packages/core/assign.js [hunk 12]
```diff
@@ -1,21 +1,18 @@
...
+
+const assignOne = (dst, src) => {
   if (src != null) {
     // We need to copy regular properties, symbols, getters and setters.
-    var keys = getOwnPropertyKeys(src);
-    for (var i = 0; i < keys.length; i += 1) {
-      var desc = Object.getOwnPropertyDescriptor(src, keys[i]);
-      Object.defineProperty(dst, keys[i], desc);
-    }
+    getOwnPropertyKeys(src).forEach((key) => defineProperty(dst, key, getOwnPropertyDescriptor(src, key)));
```
Comment (koresar) [inline packages/core/assign.js]: This `assign` is a very core function called almost every time a stamp is created, or an object is created from a stamp. Please, leave the performance optimisations intact. Namely, the `for` loops. We don't want to create extract closures. It impacts memory and CPU usage.

Same for the `reduce` below.

Sorry. I should have left more comments in the code explaining the choice.

File: packages/core/get-own-property-keys.js [hunk 13]
```diff
@@ -1,5 +1,5 @@
-module.exports = function getOwnPropertyKeys(obj) {
-  var keys = Object.getOwnPropertyNames(obj);
-  if (!Object.getOwnPropertySymbols) return keys;
-  return keys.concat(Object.getOwnPropertySymbols(obj));
-};
+'use strict';
+
+const { ownKeys: getOwnPropertyKeys } = Reflect;
```
Comment (koresar) [inline packages/core/get-own-property-keys.js]: TIL `Reflect.ownKeys`

Awesome! Thank you.

Sound like we don't need this file any more. :) We can just use `Reflect.ownKeys` everywhere! Yay!

File: packages/core/merge.js [hunk 14]
```diff
@@ -10,41 +14,40 @@ var getOwnPropertyKeys = require("./get-own-property-keys");
...
-      // Recursive calls to mergeOne() must allow only plain objects or arrays in dst
-      var newDst = isPlainObject(dst[key]) || isArray(srcValue) ? dst[key] : {};
-      // deep merge each property. Recursion!
-      dst[key] = mergeOne(newDst, srcValue);
-    } else { // nope, it looks like a getter/setter
-      Object.defineProperty(dst, key, desc);
+    // Now deal with non plain 'src' object. 'src' overrides 'dst'
+    // Note that functions are also assigned! We do not deep merge functions.
+    if (isPlainObject(src)) {
+      getOwnPropertyKeys(src).forEach((key) => {
```
Comment (koresar) [inline packages/core/merge.js]: This `merge` function was optimised by me. It is also called from many places. However, frankly speaking, not that many people use `deep...` things. :)

Please, return back the performance optimisations. This is stamp's hot place.

File: packages/it/index.js [hunk 15]
```diff
@@ -1,97 +1,99 @@
...
-  'use strict'; // to make sure `this` is not pointing to `global` or `window`
-  var length = arguments.length, args = [];
-  for (var i = 0; i < length; i += 1) {
-    var arg = arguments[i];
-    args.push(isStamp(arg) ? arg : standardiseDescriptor(arg));
-  }
-
-  return compose.apply(this || baseStampit, args); // jshint ignore:line
+function stampit(...args) {
+  return compose.apply(this || baseStampit, args.map((arg) => (isStamp(arg) ? arg : standardiseDescriptor(arg))));
```
Comment (koresar) [inline packages/it/index.js]: This was also a perf optimisation. :) But this function is not the hot area. So converting to `map()` in this case has benefits like shorter code. 👍

File: packages/named/index.js [hunk 16]
```diff
@@ -1,19 +1,18 @@
-var compose = require('@stamp/compose');
+'use strict';
```
Comment (koresar) [inline packages/named/index.js]: This module is deprecated. :) We can pretty much remove it from the repo.

**Comment by PopGoesTheWza** [comment]
> As a hint of wrongly chosen ESLint rules - see how many times you had to add `eslint-disable`. I counted 78. Wow! So much work. But what's the benefit? To satisfy airbnb?

Following my update from yesterday, I count a total of **105** `eslint-disable` occurrences.

- **99** of them are in `__test__` and `check-compose` files for which I only applied minimal editing (i.e. to pass `prettier` and `lint`.) The vast majority of these relates with `jest` best practices for what I gathered. Since unit testing is not my strong point and my knowledge of `jest`is minimal, I have not investigate further. Still, my feeling is that unit tests usually benefit from any mecanics (like linting rules) which enforce good quality unit test. We'll all seen too many projects with poorly designed unit tests which only enable the project manager to check a mark on his dashboard "unit tests: Done")
- **1** `eslint-disable-next-line no-nested-ternary` (in `packages/collision/index.js`): reasoning being that nested ternary should be used sparingly and require the developer to manually override the rule when it make sense.
- **1** `eslint-disable-next-line no-unused-vars` (in `packages/convert-class/index.js`): reasoning being that unused vars should occur for a reason and require the developer to manually override the rule when it make sense.
- **2** `eslint-disable global-require` (in `packages/core/index.js` and `packages/is/index.js`: reasoning being that `require()` should be global and preferably at the top. This rule is likely to become irrelevant in Typescript.
- **2** `eslint-disable no-use-before-define` (in `packages/it/index.js` and `packages/required/index.js`): I have not yet fully read this [rule](https://eslint.org/docs/rules/no-use-before-define) and thus decided on its relevancy.

> If you still believe that airbnb was the right choice and you would like to continue maintaining it - then I'm okay with the ESLint configuration.

I find airbnb base package to be an easy and quick way to get started, but after reading your feedback, I think it might be relevant at one point to explicitly move airbnb rules directly within `.eslintrc.json` to prevent any "blackbox" effect of the airbnb package update

**Comment by PopGoesTheWza** [comment]
> Apologies if I was somewhat grumpy.

No worries.

File: packages/compose/index.js (hunk 11 above)
Comment (PopGoesTheWza) [reply packages/compose/index.js]: Now with `for (... of ...)` inside

File: packages/core/assign.js (hunk 12 above)
Comment (PopGoesTheWza) [reply packages/core/assign.js]: Now with `for (... of ...)` inside

File: packages/core/get-own-property-keys.js (hunk 13 above)
Comment (PopGoesTheWza) [reply packages/core/get-own-property-keys.js]: Usage of `getOwnPropertyKeys()` has been changed for `Reflect.ownKeys()` directly.
The function `getOwnPropertyKeys()` is still available for `@stampit/core` but marked as `@deprecated`.

File: packages/core/merge.js (hunk 14 above)
Comment (PopGoesTheWza) [reply packages/core/merge.js]: Now with `for (... of ...)` inside

File: packages/named/index.js (hunk 16 above)
Comment (PopGoesTheWza) [reply packages/named/index.js]: Should the npm package be [deprecated](https://docs.npmjs.com/deprecating-and-undeprecating-packages-or-package-versions)?

Can you update the following files with deprecation reasoning?

- `packages/named/index.js`
- `packages/named/README.md`

File: packages/compose/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (PopGoesTheWza) [reply packages/compose/index.js]: When dropping NodeJS 8 in January, we'll set target to `es2018` (in either .eslintrc.json or typescript tsconfig.json) though there is little to benefit from it:

- Object methods `entries`, `getOwnPropertyDescriptors` and `values`
- Object.prototype getter/setter methods
- object rest/spread syntax

I have not yet fully investigate the `Proxy` object but it could find its use in a few places like `@stamp/collision` and `@stamp/privatize`

**Comment by koresar** [comment]
I've disabled all the `jest` rules for `check-compose` folder. We should be good now. :)

**Comment by koresar** [comment]
I still don't like the fact that I had to add configuration to disable other configuration. This is configuration inception. :)

**Comment by PopGoesTheWza** [comment]
That’s meta linking.

File: packages/named/index.js (hunk 16 above)
Comment (PopGoesTheWza) [reply packages/named/index.js]: @koresar ?

Comment (koresar) [reply packages/named/index.js]: Sorry. Don't know how to do that. Have no time currently. Sorry again. Life.

Comment (PopGoesTheWza) [reply packages/named/index.js]: No worries. We can do this later.
